text/3453-f16-and-f128.md

```diff
@@ -0,0 +1,124 @@
+
+## `f128` type
+
+`f128` consists of 1 bit of sign, 15 bits of exponent, 112 bits of mantissa. It is exactly equivalent to the 128-bit IEEE 754 binary128 [Quadruple-precision floating-point format](https://en.wikipedia.org/wiki/Quadruple-precision_floating-point_format).
+
+The following traits will be implemented for conversion between `f128` and other types:
+
+```rust
+impl From<f16> for f128 { /* ... */ }
+impl From<f32> for f128 { /* ... */ }
+impl From<f64> for f128 { /* ... */ }
+impl From<bool> for f128 { /* ... */ }
+impl From<u8> for f128 { /* ... */ }
+impl From<i8> for f128 { /* ... */ }
+impl From<u16> for f128 { /* ... */ }
+impl From<i16> for f128 { /* ... */ }
+impl From<u32> for f128 { /* ... */ }
+impl From<i32> for f128 { /* ... */ }
+impl From<u64> for f128 { /* ... */ }
+impl From<i64> for f128 { /* ... */ }
+```
+
+`f128` will generate `fp128` type in LLVM IR.
+
+On the hardware level, `f128` can be accelerated on RISC-V via [the Q extension](https://five-embeddev.com/riscv-isa-manual/latest/q.html), and on PowerPC with [Power ISA 3](https://en.wikipedia.org/wiki/Power_ISA#Power_ISA_v.3.0) or [later](https://fedoraproject.org/wiki/Changes/PPC64LE_Float128_Transition). Most platforms do not have hardware support and therefore will need to use a software implementation.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+While `f32` and `f64` have very broad support in most hardware, hardware support for `f16` and `f128` is more niche. On most systems software emulation will be required. Therefore, the main drawback is implementation difficulty.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+There are some crates aiming for similar functionality:
+
+- [f128](https://github.com/jkarns275/f128) provides binding to `__float128` type in GCC.
+- [half](https://github.com/starkat99/half-rs) provides implementation of binary16 and bfloat16 types.
+
+However, besides the disadvantage of usage inconsistency between primitive type and type from crate, there are still issues around those bindings.
+
+The ability to accelerate additional float types depends on CPU/OS/ABI/features of different targets heavily. Evolution of LLVM may unlock possibility of accelerating the types on new targets. Implementing them in the compiler allows the compiler to perform optimizations for hardware with native support for these types.
+
+Crates may define their type on top of a C binding, but extended float type definition in C is complex and confusing. The meaning of C types may vary by target and/or compiler options. Implementing `f16` and `f128` in the Rust compiler helps to maintain a stable codegen interface and ensures that all users have one single canonical definition of 16-bit and 128-bit float types, making it easier to exchange data between crates and languages.
+
+# Prior art
+[prior-art]: #prior-art
```

Comment:
This is a good section to put stuff about how other languages do things and what works / doesn't work well for them if anything is notable (probably not for this)

I'd mention that C++ supports these types already, and say how glibc provides a soft f128 implementation on ia64, x86_32, x86_64, and powerpc64le ([source](https://sourceware.org/git/?p=glibc.git;a=blob;f=NEWS;hb=81325b12b14c44887f1633a2c180a413afc2b504#l143)).